library/ssl_tls13_server.c

```diff
@@ -2769,9 +2779,61 @@ static int ssl_tls13_write_server_finished(mbedtls_ssl_context *ssl)
         return ret;
     }
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+    if (ssl->early_data_status == MBEDTLS_SSL_EARLY_DATA_STATUS_ACCEPTED) {
+        /* TODO: compute early transform here?
+         *
+         * RFC 8446, section A.2
+         *                      | Send Finished
+         *                      | K_send = application
+         *             +--------+--------+
+         *    No 0-RTT |                 | 0-RTT
+         *             |                 |
+         *             |                 | K_recv = early data
+         *             |    +------> WAIT_EOED -+
+         *
+         * early transform is set after server finished in this section. But
+         * it breaks our key computation, so we put early transform computation
+         * at the end of client hello. For time being, I am not sure the benifit
+         * for moving computation here.
```

Comment:
Sorry, maybe I missed some history.   Are there requirements that we should put the early data key computation and switch together? Per my understanding, it's enough that we switch to the early transform in the proper place, is it?

Reply:
See above comments, it is suggested by RFC 8446. (`*             |                 | K_recv = early data`)

Reply:
> See above comments, it is suggested by RFC 8446. (`* | | K_recv = early data`)

Seems the diagram here indicate that we switch the inbound transform to early data, but doesn't tell when to do the computation, does it?

Reply:
it has been removed. Please close this conversation if no other topic.